hexer/H3grid.hpp

```diff
@@ -0,0 +1,137 @@
+#pragma once
+
+#include <map>
+#include <vector>
+#include <sstream>
+
+#include <hexer/hexer.hpp>
+
+#include "Mathpair.hpp"
+#include "export.hpp"
+#include "H3Path.hpp"
+
+#include <h3/include/h3api.h>
+
+static bool operator<(CoordIJ const& c1, CoordIJ const& c2)
+    {
+        return (c1.i < c2.i) || ((c1.i == c2.i) && (c1.j < c2.j));
+    }
+static bool operator==(CoordIJ const& c1, CoordIJ const& c2)
+    {   
+        return ((c1.i == c2.i) && (c1.j == c2.j)); 
+    }
+
+namespace hexer 
+{
+
+using DirEdge = H3Index;
+
+static CoordIJ operator+(CoordIJ const& c1, CoordIJ const& c2)
+    {   return {c1.i + c2.i, c1.j + c2.j};  }
+
+class HEXER_DLL H3Grid
+{
+public:
+    H3Grid(int dense_limit, int resolution)
+        : m_dense_limit{dense_limit}, m_res{resolution},
+          m_maxSample{10000}, m_origin{0}
+    {}
+
+    ~H3Grid()
+    {
+        for (std::vector<H3Path*>::size_type i = 0; i < m_paths.size(); i++)
+            delete m_paths[i];
+    }
+    
+    void addLatLng(LatLng *ll);
+    std::map<CoordIJ, int> getMap() const 
+        { return m_map; }
+    int getRes() const
+        { return m_res; }
+    void setSampleSize(unsigned sampleSize)
+        { m_maxSample = sampleSize; }
+    std::vector<H3Path *> const& rootPaths() const
+        { return m_paths; }
+    int numHex()
+        { return m_numdense; }
+    // debug function, add grid for testing
+    void setGrid(std::pair<CoordIJ, int> cell)
+        { m_map.insert(cell); }
+    // debug function, specify origin for H3 grid
+    void setOrigin(H3Index idx)
+        { m_origin = idx; }
+    void processGrid();
+    void processPaths();
+
+    // Convert IJ coordinates to H3 index
+    H3Index ij2h3(CoordIJ ij)
+        {   H3Index h3;
+            if (localIjToCell(m_origin, &ij, 0, &h3) != E_SUCCESS) {
+                std::ostringstream oss;
+                oss << "Can't convert IJ (" << std::to_string(ij.i) << 
+                    ", " << std::to_string(ij.j) <<") to H3Index.";
+                throw hexer_error(oss.str()); 
+            }
+            return h3;  }
+
+    // Convert H3 index to IJ coordinates
+    CoordIJ h32ij(H3Index h3)
+        {   CoordIJ ij;
+            if (cellToLocalIj(m_origin, h3, 0, &ij) != E_SUCCESS) {
+                std::ostringstream oss;
+                oss << "Can't convert H3 index " << std::to_string(h3) << 
+                    " to IJ.";
+                throw hexer_error(oss.str()); 
+            }
+            return ij;  }
+
+    // Return the IJ coordinate of the next cell we're checking for density (going clockwise).
+    CoordIJ nextCoord(CoordIJ ij, int edge)
+        {   edge++;
+            if (edge == 6)
+                edge = 0;
+            return edgeCoord(ij, edge);  }
+
+    // Return the IJ coordinate of the cell across the edge.
+    CoordIJ edgeCoord(CoordIJ ij, int edge)
+        {   std::vector<CoordIJ> offsets {{1, 0}, {0, -1}, {-1, -1}, {-1, 0}, {0, 1}, {1, 1}};
```

Comment:
This could also be an static const array, which avoids some overhead:

```
const CoordIJ offsets[] {{1, 0}, {0, -1}, {-1, -1}, {-1, 0}, {0, 1}, {1, 1}};
```